include/unifex/stop_on_request.hpp

```diff
@@ -0,0 +1,252 @@
+  using callback_t = manual_lifetime<
+      typename StopToken::template callback_type<cancel_callback>>;
+
+  UNIFEX_NO_UNIQUE_ADDRESS callback_t<stop_token_type> receiverStopCallback_;
+  UNIFEX_NO_UNIQUE_ADDRESS std::tuple<callback_t<StopTokens>...> stopCallbacks_;
+
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+  std::atomic<_callback_state> callbackState_{_callback_state::INIT};
+
+  template(typename Receiver2)                            //
+      (requires constructible_from<Receiver, Receiver2>)  //
+      type(Receiver2&& receiver, std::tuple<StopTokens...> stopTokens)
+    : stopTokens_(std::move(stopTokens))
+    , receiver_(std::forward<Receiver2>(receiver)) {}
+
+  template <size_t index = 0>
+  constexpr void constructCallbacks() {
+    if constexpr (std::tuple_size<decltype(stopTokens_)>::value == index) {
+      // hit the end of the tuple, exit
+      return;
+    } else {
+      std::get<index>(stopCallbacks_)
+          .construct(std::get<index>(stopTokens_), cancel_callback{*this});
+      scope_guard destructOnError = [this]() noexcept {
+        // If the following callback failed to construct, then we need to clean
+        // up the current callback, eventually cleaning up all of the callbacks
+        // that have been constructed so far
+        std::get<index>(stopCallbacks_).destruct();
+      };
+
+      constructCallbacks<index + 1>();
+
+      destructOnError.release();
+    }
+  }
+
+  void start() noexcept {
+    UNIFEX_ASSERT(
+        sizeof...(StopTokens) > 0 || get_stop_token(receiver_).stop_possible());
+
+    UNIFEX_TRY {
+      // construct a callback for the receiver stop token
+      receiverStopCallback_.construct(
+          get_stop_token(receiver_), cancel_callback{*this});
+
+      scope_guard destructOnError = [this]() noexcept {
+        receiverStopCallback_.destruct();
+      };
+
+      // construct a callback for each stop token
+      // Note: if any of these tokens are already stopped, then its
+      // corresponding callback will be immediately invoked
+      constructCallbacks();
+
+      destructOnError.release();
+    }
+    UNIFEX_CATCH(...) {
+      if (callbackState_.load(std::memory_order_acquire) ==
+          _callback_state::AT_LEAST_ONE_CALLED) {
+        // we received a stop request before we experienced an error
+        unifex::set_done(std::move(receiver_));
+      } else {
+        unifex::set_error(std::move(receiver_), std::current_exception());
+      }
+
+      return;
+    }
+
+    auto expected = _callback_state::INIT;
+    // Update state to mark that all callbacks have been constructed only if the
+    // previous state was INIT. If the previous state was AT_LEAST_ONE_CALLED,
+    // don't change the state
+    if (!callbackState_.compare_exchange_strong(
+            expected,
+            _callback_state::ALL_CONSTRUCTED_NOT_CALLED,
+            std::memory_order_acq_rel)) {
+      // complete if we transitioned from AT_LEAST_ONE_CALLED
+      complete();
```

Comment:
Let's add an assert here.
```suggestion
      UNIFEX_ASSERT(expected == _callback_state::AT_LEAST_ONE_CALLED);
      complete();
```